library/src/scala/deriving.scala

```diff
@@ -0,0 +1,72 @@
+package scala
+
+object deriving {
+
+  /** Mirrors allows typelevel access to enums, case classes and objects, and their sealed parents.
+   */
+  sealed trait Mirror {
+
+    /** The mirrored *-type */
+    type MirroredMonoType
+
+    /** The name of the type */
+    type MirroredLabel <: String
+
+    /** The names of the product elements */
+    type MirroredElemLabels <: Tuple
+  }
+
+  object Mirror {
+
+    /** The Mirror for a sum type */
+    trait Sum extends Mirror { self =>
+      /** The ordinal number of the case class of `x`. For enums, `ordinal(x) == x.ordinal` */
+      def ordinal(x: MirroredMonoType): Int
+    }
+
+    /** The Mirror for a product type */
+    trait Product extends Mirror {
+
+      /** Create a new instance of type `T` with elements taken from product `p`. */
+      def fromProduct(p: scala.Product): MirroredMonoType
+    }
+
+    trait Singleton extends Product {
+      type MirroredMonoType = this.type
+      type MirroredType = this.type
+      type MirroredElemTypes = Unit
+      type MirroredElemLabels = Unit
```

Comment:
This fails when compiling from scala 2 as Unit is not a subtype of Tuple

Reply:
What's the best way to fix this? We can remove the bound in `Mirror` (if we do we need to add an `& Tuple` at one use-site in `tests/run/typeclass-derivation3.scala`)? Or we could move this file to `src-3.x`?

Reply:
Yes, we should move it to `src-3.x`. We may also need a dummy version of it in `src-2.x` the only defines classes, methods and types that we refere to in Definitions.

Reply:
Could the dummy version be identical, but without the `Tuple` bound?

Reply:
It is possible.

Reply:
PR with a fix here: https://github.com/lampepfl/dotty/pull/6609.